stratosphere/loader/source/ldr_process_creation.cpp

```diff
@@ -14,248 +14,698 @@
+            CreateProcessFlag_AddressSpace32BitWithoutAlias = (ldr::Npdm::AddressSpaceType_32BitWithoutAlias << CreateProcessFlag_AddressSpaceShift),
+            CreateProcessFlag_AddressSpace64Bit             = (ldr::Npdm::AddressSpaceType_64Bit << CreateProcessFlag_AddressSpaceShift),
+
+            /* Should JIT debug be done on crash? */
+            CreateProcessFlag_EnableDebug   = (1 << 4),
+
+            /* Should ASLR be enabled for the process? */
+            CreateProcessFlag_EnableAslr    = (1 << 5),
+
+            /* Is the process an application? */
+            CreateProcessFlag_IsApplication = (1 << 6),
+
+            /* 4.x deprecated: Should use secure memory? */
+            CreateProcessFlag_DeprecatedUseSecureMemory = (1 << 7),
+
+            /* 5.x+ Pool partition type. */
+            CreateProcessFlag_PoolPartitionShift            = 7,
+            CreateProcessFlag_PoolPartitionMask             = (0xF << CreateProcessFlag_PoolPartitionShift),
+            CreateProcessFlag_PoolPartitionApplication      = (ldr::Acid::PoolPartition_Application << CreateProcessFlag_PoolPartitionShift),
+            CreateProcessFlag_PoolPartitionApplet           = (ldr::Acid::PoolPartition_Applet << CreateProcessFlag_PoolPartitionShift),
+            CreateProcessFlag_PoolPartitionSystem           = (ldr::Acid::PoolPartition_System << CreateProcessFlag_PoolPartitionShift),
+            CreateProcessFlag_PoolPartitionSystemNonSecure  = (ldr::Acid::PoolPartition_SystemNonSecure << CreateProcessFlag_PoolPartitionShift),
+
+            /* 7.x+ Should memory allocation be optimized? This requires IsApplication. */
+            CreateProcessFlag_OptimizeMemoryAllocation = (1 << 11),
+        };
+
+    }
+
 }
 
-Result ProcessCreation::ValidateProcessVersion(u64 title_id, u32 version) {
-    if (GetRuntimeFirmwareVersion() < FirmwareVersion_810) {
-        return ResultSuccess;
-    } else {
-#ifdef LDR_VALIDATE_PROCESS_VERSION
-        if (IsDisallowedVersion810(title_id, version)) {
-            return ResultLoaderInvalidVersion;
-        } else {
+namespace sts::ldr {
+
+    namespace {
+
+        /* Convenience defines. */
+        constexpr size_t BaseAddressAlignment = 0x200000;
+        constexpr size_t SystemResourceSizeAlignment = 0x200000;
+        constexpr size_t SystemResourceSizeMax = 0x1FE00000;
+
+        /* Types. */
+        enum NsoIndex {
+            Nso_Rtld    =  0,
+            Nso_Main    =  1,
+            Nso_SubSdk0 =  2,
+            Nso_SubSdk1 =  3,
+            Nso_SubSdk2 =  4,
+            Nso_SubSdk3 =  5,
+            Nso_SubSdk4 =  6,
+            Nso_SubSdk5 =  7,
+            Nso_SubSdk6 =  8,
+            Nso_SubSdk7 =  9,
+            Nso_SubSdk8 = 10,
+            Nso_SubSdk9 = 11,
+            Nso_Sdk     = 12,
+            Nso_Count,
+        };
+
+        constexpr const char *GetNsoName(size_t idx) {
+            if (idx >= Nso_Count) {
+                std::abort();
```

Comment:
I'm not sure if it's legal to have abort in a constexpr function. This makes it impure.

Reply:
I am not sure either. It compiles, though, so it's probably fine?